### PR TITLE
roffit: fix special characters and broken links

### PR DESCRIPTION
**Pull request by jhauga**
# Pull Summary
Tested this with files matching issues this pull request resolves and with repos that have an extensive amount of files roffit can convert to html. Made two supporting repos with test results. They are:

1. File Test - roffit-test 
   - *ctrl+click* -  [RESULTS - File Test](https://jhauga.github.io/htmlpreview.github.com/?https://github.com/jhauga/roffit-test/blob/main/index.html)
   - *ctrl+click* -  [roffit-test repository](https://github.com/jhauga/roffit-test)
2. Repo Test - roffit-full-test
   - *ctrl+click* - [RESULTS - Repo Test](https://jhauga.github.io/htmlpreview.github.com/?https://github.com/jhauga/roffit-full-test/blob/main/index.html)
   - *ctrl+click* - [roffit-full-test repository](https://github.com/jhauga/roffit-full-test)

The two tests used more or less the same procedure, and focused on:
1. Files with differences
2. Build time

The two support repos have full details of the procedure, but in short the steps were:
1. Extract data and ready test elements
3. Clone source (*bagder*) and pull (*jhauga*) roffit repo.
4. Run roffit and time each runs' build time.
5. Determine files with differences.
6. Calculate stats for files with difference and the build times. 
7. Output results to html file.

### CODE REVIEW

**Review by jhauga — COMMENTED** [review]
Tested build on local host. 4 make rules generated with no problem. For more details see the README.md in "examples_DELETE_ME" folder.

**Review by bagder — CHANGES_REQUESTED** [review]
I don't see any actual roffit changes here? Does it only add tests? How can it then solve any problems?

Also, this has to be possible to test with much less data added. 23K of data is too excessive.

**Comment by jhauga** [comment]
> \***1** I don't see any actual roffit changes here? 
> \***2** Does it only add tests? 
> \***3** How can it then solve any problems?
> \***4** 23K of data is too excessive.

First of all I'd like to thank you and the rest of the curl team for 
curl and debatably the contemporary internet. 

Next I'll go ahead and disclose that I'm at an elementary level in 
terms of using GitHub. With that said to answer your questions:

1. Yes - this would be VERY hard (*i think*) to accomplish in roffit
2. Yes - 2 lines in the Makefile that call different bash script files
3. The scripts resolve 2 cases that are not easily resolvable in 
roffit by itself:
   - **A** Broken links caused by inline and quoted full urls in ".1" 
   file.
      - fixBrokenLinks.sh - uses sed in conjunction with grep and cut 
	  to solve this.
   - **B** Special characters being excluded in links and anchor links. 
      - fixSpecialCharacters.sh - prompts (*config variable may be better*)
	  after issuing a make command, asking if special characters used in options;
4. Will do, but the example folder provided is meant to be a proof of concept for 
   Make file with 2 scripts, and deleted. It was built and tested outside of Github
   with php local host.
   For example php -S localhost:XXXX; then from that either:
   - 2 variations of full curl man page are built using Makefile with scripts
      - 1 as built with curl the other with 3 ".d" options files edited,
	  which resolves unwanted inline links. (*ftp-port, proto, page-footer*).
   - 2 condensed curl man pages with only options and section mentioned in
   [curl #11381](https://github.com/curl/curl/issues/11381).
      - 1 as built with curl the other with 3 ".d" options files edited,
	  which resolves unwanted inline links. (*ftp-port, proto, page-footer*).

**Comment by bagder** [comment]
I approve of extending the tests to test for more problems, but I think this PR goes about it the wrong way.

A first take would be to extend the existing test case with more text for the problems you have identified. Not adding thousands of lines of new files.

roffit is not curl specific, it converts man pages to html so it has no knowledge of or special handling for curl related issues.

**Comment by jhauga** [comment]
> I approve of extending the tests to test for more problems, but I think this PR goes about it the wrong way.
> 
> A first take would be to extend the existing test case with more text for the problems you have identified. Not adding thousands of lines of new files.
> 
> roffit is not curl specific, it converts man pages to html so it has no knowledge of or special handling for curl related issues.

Sounds good. I'm working on cleaning things up, and will remove the excessive test folder in next pull request. 
Tinkering with roffit a bit more this time and may edit it; so I should hopefully have this ready in the next few days. 
Definitely no later than next Friday though.

**Comment by jhauga** [comment]
> I approve of extending the tests to test for more problems, but I think this PR goes about it the wrong way.
> 
> A first take would be to extend the existing test case with more text for the problems you have identified. Not adding thousands of lines of new files.
> 
> roffit is not curl specific, it converts man pages to html so it has no knowledge of or special handling for curl related issues.

## Requested Changes
Corrected title and made requested changes.

- Deleted excessive test folder.
- Added examples in test page that include items this pull request fixes. 

## Additional Edits
- Made edits in roffit file similar to [roffit #39](https://github.com/bagder/roffit/pull/39)  that would have reopen [roffit #38](https://github.com/bagder/roffit/issues/38).  
- Edited supporting script files fixBrokenLinks and fixSpecialCharacters. 
- Edited Makefile adding short descriptive comments for supporting script files.

### Additional note
In the next several months I may open another pull request to resolve issue #21. It has been open since 2019 so I thought I'd note that a variation of fixSpecialCharacters.sh (* i.e. addDoubleOptionAnchor.sh *) may resolve [roffit 21](https://github.com/bagder/roffit/issues/21), where a rule or condition in Makefile would add anchors for both options. If I get a solid foundation in perl, then tinker with roffit file some more.

File: roffit
Comment (jhauga) [inline roffit:1]: for 111 and 155 see [xhtml standard](https://www.w3.org/TR/xhtml1/#h-4.2)

File: fixSpecialCharacters.sh
Comment (jhauga) [inline fixSpecialCharacters.sh:1]: Some man pages with special characters include:
less, tar, gzip, unzip, mysql, traceroute, and rsync.

File: fixBrokenLinks.sh
Comment (jhauga) [inline fixBrokenLinks.sh:1]: A bit embarrassed, but did not change the file name back to testpage.dump in the original pull request. 
This is my first real pull request so a bit nervous I guess. Regardless, this has been a very rewarding process.

File: examples_DELETE_ME/fixBrokenLinks.sh
Comment (jhauga) [inline examples_DELETE_ME/fixBrokenLinks.sh:1]: This script (*and other .sh script *) was same as script  in root folder, but the file name was not changed back to testpage.dump when pasted back to root. Embarrassed but thought I'd mention so there is no confusion. Happy changes were requested!

File: Makefile
Comment (jhauga) [inline Makefile:1]: If fixBrokenLinks is called before fixSpecialCharacters, then special characters are not converted conrrectly. They become "#" top of page anchors.

**Review by jhauga — COMMENTED** [review]
Reviewed with comments on Makefile, fixBrokenLink (*edited and deleted*), fixSpecialCharacters, and roffit files.

**Comment by bagder** [comment]
Sorry, but this PR does not fix roffit as described. It just fixes two minor static HTML texts.

**Comment by jhauga** [comment]
> Sorry, but this PR does not fix roffit as described. It just fixes two minor static HTML texts.

That's fine. Thank you for considering it. Adding two bash scripts was the only way I was able to resolve [issue 36](https://github.com/bagder/roffit/issues/36). Learning more about perl is on my todo list, so if the issue is still up in several months I may try again.

Regardless, this has been a very enjoyable and rewarding process!

**Comment by bagder** [comment]
Separate scripts cannot solve an issue. `roffit` is the program that runs that converts nroff files, so you need to fix that file. Having separate shell scripts is not a solution to the problem!

**Comment by jhauga** [comment]
> Separate scripts cannot solve this issue. `roffit` is the program that runs that converts nroff files, so you need to fix that file. Having separate shell scripts is not a solution to the problem!

Ok. I didn't know. I thought calling the scripts in the Makefile after roffit ran could be part of the make recipe. They did work though. 

I will have become more fluent in perl before I'm able to implement this in roffit.

**Comment by jhauga** [comment]
Reopening was able to fix in roffit file without (*kinda*) bash scripts.

**Review by bagder — CHANGES_REQUESTED** [review]
This is totally unacceptable code. Make it clean pure perl and do not generate temp files.

**Comment by jhauga** [comment]
> This is totally unacceptable code. Make it clean pure perl and do not generate temp files.

Will do. If not able to achieve by Friday will close and then reopen again when/if able to achieve as requested.

**Review by jhauga — COMMENTED** [review]
lines 259 to 285 - perl changes

File: roffit [hunk 1]
```diff
@@ -256,6 +256,33 @@ sub linkfile {
             # convert https://, http:// and ftp:// URLs to <a href> links
             $field =~ s/(^|\W)((https|http|ftp):\/\/[a-z0-9\-._~%:\/?\#\[\]\@!\$&'()*+,;=]+)/$1<a href=\"$2\">$2<\/a>/gi;
 
+            # remove punctuation characters at end of url
+            $field =~ s/(href=\")([^\"]*)(&quot;)/$1$2/g;
```
Comment (bagder) [inline roffit:260]: This does not remove punctuation. What is this for?

File: roffit [hunk 2]
```diff
@@ -256,6 +256,33 @@ sub linkfile {
...
+            $field =~ s/(href=\"(http:\/\/|https:\/\/|ftp:\/\/)\")*([,.;()*])*(\">)/$1$4/g;
```
Comment (bagder) [inline roffit:261]: Shouldn't this rather correct the pattern on line 257 to not include punctuation  as terminating symbols?

File: roffit [hunk 3]
```diff
@@ -256,6 +256,33 @@ sub linkfile {
...
+
+            # null empty protocols
+            $field =~ s/href=\"(http:\/\/|https:\/\/|ftp:\/\/)\"/href=\"javascript:void(0)\"/g;
```
Comment (bagder) [inline roffit:264]: If the above is fixed, then this kind of link should not get generated. Right? And I would prefer not having this a link at all rather than to have this javascript-thing. In fact, roffit generates HTML (and CSS) only, I don't want to add javascript calls to the mix.

File: roffit [hunk 4]
```diff
@@ -256,6 +256,33 @@ sub linkfile {
...
+
+            # add anchor and anchor links to special character options
+            my $specialcharacters = "`~!@\$%^*()-_=+{};:\'\\|,.?";
+            $field =~ s/(<a name=)\"-\"(><\/a>)(<span class=\"nroffip\">-)([$specialcharacters]|[\[\]])(,|<\/span>)/$1\"-$4\"$2$3$4/g;
+            $field =~ s/(<a class=\"emphasis\") href=\"#-\">-([$specialcharacters]|[\[\]])(,|<\/a>)/$1 href=\"#-$2\">-$2/g;
```
Comment (bagder) [inline roffit:269]: I think these changes should rather be done as updates to `text2name()` and `do_encode()` instead of adding this extra stage of edits.

File: roffit [hunk 5]
```diff
@@ -256,6 +256,33 @@ sub linkfile {
...
+
+            # add anchor to html entities
+            my $htmlentity = "(&#35;)|(&amp;)|(&#39;)|(&lt;)|(&gt;)";
+            if ($field =~ /(<a name=)\"-\"(><\/a>)(<span class=\"nroffip\">-)($htmlentity)(,|<\/span>)/) {
+                $field =~ s/(<a name=)\"-\"(><\/a>)(<span class=\"nroffip\">-)($htmlentity)(,|<\/span>)/$1\"-$4\"$2$3$4/g;
+                $field =~ s/&#35;/hash/ | $field =~ s/&amp;/ampersand/ | $field =~ s/&#39;/single-quote/ |
+                $field =~ s/&lt;/less-than/ | $field =~ s/&gt;/greater-than/;
+            }
```
Comment (bagder) [inline roffit:277]: I think these ones could also be included into that change.

File: roffit [hunk 6]
```diff
@@ -152,7 +152,7 @@ p.roffit {
  text-align: center;
  font-size: 80%;
 }
-</STYLE>
+</style>
```
Comment (bagder) [inline roffit:155]: These are off-topic changes. I think you should do them separately.

File: testpage.1 [hunk 7]
```diff
@@ -41,9 +62,15 @@ name in the generated link will be prefixed by the dir given with
...
 foformatting) and in a plain \.BR section (often used in the SEE ALSO
 section).
+
 .IP \-\-hrefdir=<dir>
 Specify a directory to prefix generated href links created with the \-\-mandir
 option. This defaults to ".".
+
+.IP "-?, --special-char"
+Options with special characters will be included in anchors.
+Such as this option with special character \fI\-?, \-\-special\-char\fP.
```
Comment (bagder) [inline testpage.1:72]: I don't see any code handling this new option. Plus, I don't think this needs to be managed by an option - I think it makes sense to improve the general handling of anchors and links so it should probably be used by default.

File: testpage.1 [hunk 8]
```diff
@@ -65,6 +92,6 @@ Text marked as a reference to another man page.
 .IP span.emphasis
 Text marked to be emphasized.
 .IP p.roffit
-Used for the advertising final paragraph.
+Used for advertising final paragraph.
 .SH WWW
-http://daniel.haxx.se/projects/roffit
+ http://daniel.haxx.se/projects/roffit
```
Comment (bagder) [inline testpage.1:97]: Why adding a space here?

File: roffit (hunk 1 above)
Comment (jhauga) [reply roffit]: \&quot; seemed to be a troublemaker so for following lines to work it had to be removed first in process.

File: roffit (hunk 2 above)
Comment (jhauga) [reply roffit]: Tried with variations of pattern. Either my corrections were flawed or the URLs need to be built first.

File: roffit (hunk 4 above)
Comment (jhauga) [reply roffit:282]: Unable to achieve in subroutines, either anchor links were removed, all anchors (*name attribute*) removed, or no effect.

File: roffit (hunk 5 above)
Comment (jhauga) [reply roffit:288]: Created subroutine field_anchor to handle repeating blocks. Required $field and $htmlentity to be defined globally.

File: testpage.1 (hunk 7 above)
Comment (jhauga) [reply testpage.1]: For the test build. Moved into description. Anyway options are named wrong (*--mumbo and --jumbo*), which gave me impression that it was only for build test purposes. Does not seem like an issue though (*maybe good for someone to find who has never contributed on github also*) .

File: testpage.1 (hunk 8 above)
Comment (jhauga) [reply testpage.1]: I don't know. I missed that.

**Comment by jhauga** [comment]
### Regarding some suggested edits:
Unable to achieve fix with **text2name** and **do_encode**. 
In short - new things break. 

Quick takeaway ( *where* - (*1*) should be =\> (*2*) becomes ):
name="--any-option" =\> name=""--any-option" (*"" at start*)
name="ANY_NAMES" =\> name="ANYNAMES" (*_ is removed*)

### For details see:
<details>

<summary>collapsed section</summary>

### Detailed report (explains above):
___
Since curl.1 is a good use case for thoroughly testing a roffit 
build I made two files, running "diff" command on them: 
   1. with current badger roffit
   2. with (multiple) edited roffit

Below are sample outcomes from two instances of build, and then results
from using current edits.

**do_encode --- example case removes \":**

```
sub do_encode($) {
    return encode_entities(shift, q{<>&'#});
}
```
results in:
name=""--any-option" (*2 "" at start*)

**text2name --- example case adds special characters to \$text:**
```
sub text2name { .....
   $text =~ s/[^a-zA-Z0-9-`~!@\$%^*()-_=+{};:\'\\|,.?]//g;
   ...}
```
results in:
name="socks5h://" (*should be name=\"socks5h\"*).
**or**
name="AUNDERSCORE" (*shoud be name=\"A_UNDERSCORE*\").

Several variations of the above edits were done with similar (*if not the same*) outcome. 

So in conclusion; either my approaches were flawed, or I need to study the roffit 
file some more.

### When "diff" ran using  -- 2. with current pull roffit --  
- roffit built as expected, nothing new broken
- Outputted difference resolved [roffit #36](https://github.com/bagder/roffit/issues/36).
___
</details>

### Edits
- Reordered htmlentity variable to be same order as call in **do_encode** (*excluding \"*).

- Edited regular expression for special character options to fix a missed bug 
that removed inline comma in option text. (*i.e.* \"-? --opt\" =\> \"-?, --opt\").

- Removed condition in **field_anchor** subroutine to reduce line numbers, 
  moving line needed to line number 289.

File: roffit (hunk 4 above)
Comment (jhauga) [reply roffit]: > I think these changes should rather be done as updates to `text2name()` and `do_encode()` instead of adding this extra stage of edits.

With `text2name` and `do_encode` new things break. Some examples:
- name="socks5h://" (*should be name=\"socks5h\"*)
- name="AUNDERSCORE" (*shoud be name=\"A_UNDERSCORE*\")
- name=""--any-option" (*2 "" at start*).

File: roffit (hunk 2 above)
Comment (jhauga) [reply roffit]: > Shouldn't this rather correct the pattern on line 257 to not include punctuation as terminating symbols?

Kept getting unexpected results like `<a href="https://curl">https://curl</a>.se/,` or inline punctuations being removed.

File: roffit (hunk 6 above)
Comment (jhauga) [reply roffit]: Ok but if @1268 wants to do them he should. Will give it a week or so and if not resolved, then will re-open prior pull \#40.

**Comment by jhauga** [comment]
Tested this with files matching issues this pull request resolves and with repos that have an extensive amount of files roffit can convert to html. Made two supporting repos with test results. They are:

1. File Test - roffit-test 
   - *ctrl+click* -  [RESULTS - File Test](https://jhauga.github.io/htmlpreview.github.com/?https://github.com/jhauga/roffit-test/blob/main/index.html)
   - *ctrl+click* -  [roffit-test repository](https://github.com/jhauga/roffit-test)
2. Repo Test - roffit-full-test
   - *ctrl+click* - [RESULTS - Repo Test](https://jhauga.github.io/htmlpreview.github.com/?https://github.com/jhauga/roffit-full-test/blob/main/index.html)
   - *ctrl+click* - [roffit-full-test repository](https://github.com/jhauga/roffit-full-test)

The two tests used more or less the same procedure, and focused on:
1. Files with differences
2. Build time

The two support repos have full details of the procedure, but in short the steps were:
1. Extract data and ready test elements
3. Clone source (*bagder*) and pull (*jhauga*) roffit repo.
4. Run roffit and time each runs' build time.
5. Determine files with differences.
6. Calculate stats for files with difference and the build times. 
7. Output results to html file.
